docs/book/v2/migration/preparing-for-v3.md

```diff
@@ -44,3 +44,13 @@ This deprecation can be safely ignored but in order to prepare for its removal i
 In version 2, the `TemplatePathStack` template resolver automatically registers a stream wrapper for templates when the php.ini setting `short_open_tag` was turned off. The purpose of the stream wrapper was to convert template files using the short open tag `<?= $variable ?>` to `<?php echo $variable ?>` so that templates would continue to be processed in environments where short_open_tag was turned off. Since PHP 5.4.0, `<?=` is always available, therefore the wrapper became mostly unnecessary.
 
 The impact of this future removal will affect templates that use a regular short open tag for general PHP code, i.e. `<? $i = 1; echo $i ?>` in environments where `short_open_tag` is **off**. To mitigate the impact of this removal, you should ensure that, where relevant, all of your templates use the full `<?php` open tag. Use of the short echo tag `<?=` is unaffected.
+
+## View Helper Changes
+
+### [Json View Helper](../helpers/json.md)
+
+In version 3, the JSON view helper will no longer set response headers _(For MVC requests)_ when used.
+If you are using this feature, you will need to refactor your controllers to return the correct mime-type rather than relying on the view helper to do it for you.
```

Comment:
Can we add a code example here?

Reply:
I might need some help here as it's been a very long time since I used MVC. I'm guessing something like this would suffice:

```php
class MyController extends AbstractActionController {
    public function someAction(): void
    {
        // ...
        $this->response->setHeader('Content-Type', 'application/json');
        // ...
    }
}
```

I don't believe this will be a BC break for MVC. AFAICT, MVC uses the `JSONRenderingStrategy` - I couldn't find any references to this view helper in the MVC codebase. Perhaps @Xerkus can confirm.

Reply:
```php
final class AlbumController extends AbstractActionController
{
    public function listAction(): array
    {
        // …

        $this->getResponse()->getHeaders()->addHeaderLine(
            'Content-Type',
            'application/json'
        );

        return [
            'albums' => ['…']
        ];
    }
}
```

Reply:
MVC controllers do not handle content type response headers (users might do whatever, so not a given). It is done by view strategies.  Json strategy sets content type for this. Those are being extracted into laminas-mvc-view I believe.

ext-json was moved to php core extensions with php 8 and always available. This helper is obsolete imo. 

I do not know what use is there left for this helper. May be set default json_encode flags? I would suggest deprecating it and drop for v3 instead.

Via laminas-json there is a special handling of objects with `toJson()` or `toArray()` methods. It is not documented or mentioned here. Expressions are mentioned in docs but declared deprecated.


Also, this was introduced in zf1 long long time ago: https://github.com/zendframework/zf1/blob/release-1.12.0/library/Zend/View/Helper/Json.php#L85

Reply:
Actually, json response strategy is used with json view model. This helper is used with default strategy when array needs to be prepared by .phtml before getting emitted. A hack, really, because json strategy does not give controller specific hook points to prepare view model data for json format.

A mess any way you look at it.

Reply:
I think @Xerkus is correct and we should remove this helper. I see no reason why a view script should be used if a JSON output is to be created.
And the idea was to remove the `JsonStrategy` and `FeedStrategy` with their related renderers. (see #149)
But so far we have no user-friendly replacement like the [`JsonResponse` or `XmlResponse` of laminas-diactoros](https://docs.laminas.dev/laminas-diactoros/v3/custom-responses/).